### ISSUE TITLE
After auth, TCP connections fail.

### ISSUE THREAD

**Issue by alexfornuto** [pomerium/pomerium#2590]
## What happened?

On every fresh connection to a TCP route, the connection fails and needs to be restarted. By "fresh", I mean that the JWT is not cached. The connection must be closed and reopened to get a solid connection. Removing the cache re-creates the issue on the next attempt.

## What did you expect to happen?

I expected The first connection to open my browser and authenticate me, and then proceed to allow connections.

## How'd it happen?

1. `rm -r ~/.cache/pomerium-cli`
2. `❯ pomerium-cli tcp ssh.mydomain.tld:22 --listen :2202`
3. Output across first and second connection attempts:

SSH to Pomerium 15.0 (upgraded to 15.2 with no change):

  ```bash
  ❯ pomerium-cli tcp ssh.mydomain.tld:22 --listen :2202
  1:49PM INF tcptunnel: listening on [::]:2202
  1:49PM INF tcptunnel: opening connection dst=ssh.mydomain.tld:22 proxy=ssh.mydomain.tld:443 secure=true
  1:49PM INF tcptunnel: opening connection dst=ssh.mydomain.tld:22 proxy=ssh.mydomain.tld:443 secure=true
  1:49PM INF tcptunnel: connection closed
  1:49PM INF tcptunnel: connection closed
  1:49PM ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
  1:49PM INF tcptunnel: opening connection dst=ssh.mydomain.tld:22 proxy=ssh.mydomain.tld:443 secure=true
  1:49PM INF tcptunnel: connection closed
  1:49PM ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
  1:49PM INF tcptunnel: opening connection dst=ssh.mydomain.tld:22 proxy=ssh.mydomain.tld:443 secure=true
  1:49PM INF tcptunnel: connection closed
  1:49PM ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
  ^C%                                                                                                                            
  ❯ pomerium-cli tcp ssh.mydomain.tld:22 --listen :2202
  1:49PM INF tcptunnel: listening on [::]:2202
  1:49PM INF tcptunnel: opening connection dst=ssh.mydomain.tld:22 proxy=ssh.mydomain.tld:443 secure=true
  1:49PM INF tcptunnel: connection established
  ```

RDP to Pomerium 15.1

```bash
❯ pomerium-cli tcp rdp.mydomain.tld:102 --listen :3389
1:47PM INF tcptunnel: listening on [::]:3389
1:47PM INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
1:47PM INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
1:47PM INF tcptunnel: connection closed
1:47PM INF tcptunnel: connection closed
1:47PM ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
1:47PM INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
1:47PM INF tcptunnel: connection closed
1:47PM ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
^C%
❯ pomerium-cli tcp rdp.mydomain.tld:102 --listen :3389
1:48PM INF tcptunnel: listening on [::]:3389
1:48PM INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
1:48PM INF tcptunnel: connection established
```

## What's your environment like?

- Pomerium version: Present in 15.0 through 15.2
 Newer 
- Server Operating System/Architecture/Cloud: All-in-One package on Debian servers

## What did you see in the logs?

```logs
Sep 10 16:31:41 SERVER-HOSTNAME pomerium[30807]: {"level":"info","service":"envoy","name":"main","time":"2021-09-10T16:31:41-05:00","message":"shutting down parent after drain"}
Sep 10 16:31:47 SERVER-HOSTNAME pomerium[30807]: {"level":"info","service":"authorize","request-id":"b0fa0922-02a4-4ef7-bc0e-c808462c7cf0","check-request-id":"1bdd0da1-c406-4253-a260-6490686f853a","method":"CONNECT","path":"","host":"ssh.mydomain.tld:22","query":"","allow":false,"deny":null,"user":"","email":"","databroker_server_version":1891646788856662955,"databroker_record_version":2099,"time":"2021-09-10T16:31:47-05:00","message":"authorize check"}
Sep 10 16:31:47 SERVER-HOSTNAME pomerium[30807]: {"level":"info","request-id":"b0fa0922-02a4-4ef7-bc0e-c808462c7cf0","@type":"type.googleapis.com/pomerium.crypt.SealedMessage","key_id":"REDACTED","time":"2021-09-10T16:31:47-05:00","message":"audit log"}
Sep 10 16:31:48 SERVER-HOSTNAME pomerium[30807]: {"level":"info","peer":"127.0.0.1:43498","type":"type.googleapis.com/session.Session","id":"52144649-914d-4c41-818f-8ea891455f81","time":"2021-09-10T16:31:48-05:00","message":"get"}
Sep 10 16:31:48 SERVER-HOSTNAME pomerium[30807]: {"level":"info","service":"envoy","upstream-cluster":"","method":"CONNECT","authority":"ssh.mydomain.tld:22","path":"","user-agent":"Go-http-client/1.1","referer":"","forwarded-for":"$MY_IP","request-id":"1bdd0da1-c406-4253-a260-6490686f853a","duration":4.687338,"size":10665,"response-code":302,"response-code-details":"ext_authz_denied","time":"2021-09-10T16:31:48-05:00","message":"http-request"}
Sep 10 16:31:48 SERVER-HOSTNAME pomerium[30807]: {"level":"info","service":"envoy","upstream-cluster":"pomerium-control-plane-http","method":"GET","authority":"ssh.mydomain.tld","path":"/.pomerium/api/v1/login","user-agent":"Go-http-client/2.0","referer":"","forwarded-for":"$MY_IP","request-id":"c0cdd804-506c-4e7e-9bf5-5a4a4338bc10","duration":1.03364,"size":380,"response-code":200,"response-code-details":"via_upstream","time":"2021-09-10T16:31:48-05:00","message":"http-request"}
Sep 10 16:31:48 SERVER-HOSTNAME pomerium[30807]: {"level":"info","service":"envoy","upstream-cluster":"pomerium-control-plane-http","method":"GET","authority":"pom.mydomain.tld","path":"/.pomerium/sign_in","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.63 Safari/537.36","referer":"","forwarded-for":"173.29.66.14","request-id":"de26cdaa-d145-4b06-912a-76d9e40588a1","duration":3.691519,"size":916,"response-code":302,"response-code-details":"via_upstream","time":"2021-09-10T16:31:48-05:00","message":"http-request"}
Sep 10 16:31:49 SERVER-HOSTNAME pomerium[30807]: {"level":"info","service":"envoy","upstream-cluster":"pomerium-control-plane-http","method":"GET","authority":"ssh.mydomain.tld","path":"/.pomerium/callback/","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.63 Safari/537.36","referer":"","forwarded-for":"$MY_IP","request-id":"ea5ccd08-6c9d-4ebd-8c20-8db81f91032f","duration":0.9932,"size":467,"response-code":302,"response-code-details":"via_upstream","time":"2021-09-10T16:31:49-05:00","message":"http-request"}
```

**Comment by calebdoxsey** [comment]
@alexfornuto do you want to try master? I just merged a change that should fix some of these issues I think

**Comment by alexfornuto** [comment]
It was suggested during the meeting that this issue may have been related to removing the cached tokens to simulate a new connection. I can confirm that the issue happens with a fresh connection, not one with removed tokens:

```
❯ pomerium-cli tcp vnc.mydomain.tld:5888 --listen :5888
2:56PM INF tcptunnel: listening on [::]:5888
2:56PM INF tcptunnel: opening connection dst=vnc.mydomain.tld:5888 proxy=vnc.mydomain.tld:443 secure=true
2:56PM INF tcptunnel: connection closed
2:56PM ERR tcptunnel: error serving local connection error="tcptunnel: invalid http response code: 503"
# Then I added a policy to allow access
2:57PM INF tcptunnel: opening connection dst=vnc.mydomain.tld:5888 proxy=vnc.mydomain.tld:443 secure=true
2:57PM INF tcptunnel: opening connection dst=vnc.mydomain.tld:5888 proxy=vnc.mydomain.tld:443 secure=true
2:57PM INF tcptunnel: connection closed
2:57PM INF tcptunnel: connection closed
2:57PM ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
2:57PM INF tcptunnel: opening connection dst=vnc.mydomain.tld:5888 proxy=vnc.mydomain.tld:443 secure=true
2:57PM INF tcptunnel: connection closed
2:57PM ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
^C%
❯ pomerium-cli tcp vnc.mydomain.tld:5888 --listen :5888
2:57PM INF tcptunnel: listening on [::]:5888
2:57PM INF tcptunnel: opening connection dst=vnc.mydomain.tld:5888 proxy=vnc.mydomain.tld:443 secure=true
2:57PM INF tcptunnel: connection established
```



> @alexfornuto do you want to try master? I just merged a change that should fix some of these issues I think

Sure. To confirm, this is a new master build of `pomerium`  you're referring to? Do I need to update the server, local client, or both?

**Comment by calebdoxsey** [comment]
The local client

**Comment by alexfornuto** [comment]
Thanks. Unfortunately It didn't change the behavior:

```
❯ git show
commit 77ae17d23b43f143db6d3e1447f5eb5b0311c872 (HEAD -> master, origin/master, origin/HEAD)
Author: Caleb Doxsey <cdoxsey@pomerium.com>
Date:   Mon Sep 13 13:53:19 2021 -0600

...

❯ rm -r ~/.cache/pomerium-cli
❯ pwd
/home/alex/projects/pomerium
❯ go run ./cmd/pomerium-cli tcp rdp.mydomain.tld:102 --listen :3389
2021-09-13T15:12:42-05:00 INF tcptunnel: listening on [::]:3389
2021-09-13T15:12:45-05:00 INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
2021-09-13T15:12:46-05:00 INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
2021-09-13T15:12:47-05:00 INF tcptunnel: connection closed
2021-09-13T15:12:47-05:00 INF tcptunnel: connection closed
2021-09-13T15:12:47-05:00 ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
2021-09-13T15:12:47-05:00 INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
2021-09-13T15:12:48-05:00 INF tcptunnel: connection closed
2021-09-13T15:12:48-05:00 ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
2021-09-13T15:12:55-05:00 INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
2021-09-13T15:12:56-05:00 INF tcptunnel: connection closed
2021-09-13T15:12:56-05:00 ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
2021-09-13T15:12:56-05:00 INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
2021-09-13T15:12:58-05:00 INF tcptunnel: connection closed
2021-09-13T15:12:58-05:00 ERR tcptunnel: error serving local connection error="tcptunnel: failed to read HTTP response: unexpected EOF"
^C%                                                                                                                            ❯ pwd
❯ go run ./cmd/pomerium-cli tcp rdp.mydomain.tld:102 --listen :3389
2021-09-13T15:13:14-05:00 INF tcptunnel: listening on [::]:3389
2021-09-13T15:13:16-05:00 INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
2021-09-13T15:13:16-05:00 INF tcptunnel: connection established
```

**Comment by calebdoxsey** [comment]
@alexfornuto can you set `log_level` to `debug` in the server config, and grab the logs when you try to make the connection?

**Comment by alexfornuto** [comment]
I can't at the moment, as I can no longer get to that point as discussed [here](https://github.com/pomerium/pomerium/issues/2589#issuecomment-918551286). Once I can establish the CLI connection again, I can produce relevant debug output for this issue.

**Comment by alexfornuto** [comment]
I sent you the debug output over Slack, since I'm not sure how to confirm it's sanitized.

**Comment by calebdoxsey** [comment]
@alexfornuto does the latest master fix this issue? One with the changes from #2594

**Comment by alexfornuto** [comment]
Just tested, bug resolved!

```❯ rm -rf ~/.cache/pomerium-cli
❯ go run ./cmd/pomerium-cli tcp rdp.mydomain.tld:102 --listen :3389
2021-09-15T10:33:13-05:00 INF tcptunnel: listening on [::]:3389
2021-09-15T10:33:15-05:00 INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
2021-09-15T10:33:20-05:00 INF tcptunnel: opening connection dst=rdp.mydomain.tld:102 proxy=rdp.mydomain.tld:443 secure=true
2021-09-15T10:33:20-05:00 INF tcptunnel: connection established
```

Thanks @calebdoxsey!
